patches/minecraft/net/minecraft/inventory/container/RecipeBookContainer.java.patch

```diff
@@ -1,11 +1,73 @@
 --- a/net/minecraft/inventory/container/RecipeBookContainer.java
 +++ b/net/minecraft/inventory/container/RecipeBookContainer.java
-@@ -31,4 +31,8 @@
+@@ -31,4 +31,70 @@
  
     @OnlyIn(Dist.CLIENT)
     public abstract int func_203721_h();
 +
-+   public java.util.List<net.minecraft.client.util.RecipeBookCategories> getRecipeBookCategories() {
-+      return net.minecraft.client.util.ClientRecipeBook.func_216769_b(this);
++   /*================================ FORGE START ================================================*/
++   public net.minecraft.item.crafting.IRecipeType<?> getRecipeType() {
++      if (this instanceof FurnaceContainer) {
++         return net.minecraft.item.crafting.IRecipeType.field_222150_b;
++      } else if (this instanceof SmokerContainer) {
++         return net.minecraft.item.crafting.IRecipeType.field_222152_d;
++      } else if (this instanceof BlastFurnaceContainer) {
++         return net.minecraft.item.crafting.IRecipeType.field_222151_c;
++      } else if (this instanceof PlayerContainer || this instanceof WorkbenchContainer) {
++         return net.minecraft.item.crafting.IRecipeType.field_222149_a;
++      } else {
++         throw new IllegalStateException("Unexpected RecipeBookContainer override, please override this method in your own class!");
++      }
 +   }
++
++   public java.util.List<net.minecraftforge.common.extensions.IForgeRecipeBookCategory<?>> getRecipeBookCategories() {
++      return net.minecraftforge.registries.ForgeRegistries.RECIPE_BOOK_CATEGORIES.getValues().stream().filter(category -> {
++         return category.getRecipeType().equals(this.getRecipeType());
++      }).sorted(RecipeBookContainer::compareTo).collect(java.util.stream.Collectors.toList());
++   }
++
++   private static final java.util.Map<net.minecraftforge.common.extensions.IForgeRecipeBookCategory<?>, Integer> VANILLA_LIST  = new java.util.HashMap<>();
++
++   private static int compareTo(net.minecraftforge.common.extensions.IForgeRecipeBookCategory<?> from, net.minecraftforge.common.extensions.IForgeRecipeBookCategory<?> to) {
```

Comment:
Ordering is definitely needed there, first I must provide correctly ordered vanilla categories and secondly to not break vanilla ordering, I have to provide as first category "search".

Other categories are then in registration order

Reply:
Since setup events are fired concurrently, wouldn't that make the category order for mod categories indeterministic? It might be a better user experience to sort mod categories by name

Reply:
Setup events are not parallel, meaning this somewhat guarantees ordering. They may become parallel in the future, though.
But at the same time I don't see how this is an issue, considering the same happens for item groups, blocks, items themselves... There is a lot of reliance on registration order so I believe this is fine.

Reply:
@yanny7 What is the exact reasoning for the ordering? Does vanilla do an index based approach here and just assumes that "Search" is always index 0?

Or why does Vanilla want these pages to be first?

Reply:
In vanilla order is hard-coded, but always in first tab are recipes without filtering, what makes sense. I just wanna make it consistent.

I named it wrong "search", because you can search in every tab. But as I mentioned, first tab is unfiltered list of recipes

Reply:
But where is it hard-coded? Is it just the way they add the tabs to the list? Are they accessing that list then again to get a tab back an rely on the unfiltered list being on 0?

Cause at the moment I do not see a reason for the ordering except for: Vanilla adds it this way, so that is the reason.

Do not get me wrong. Having ordering on something like this would be nice since it produces some consistency in a much used module.

Reply:
Please check old implementation:
```
public static List<RecipeBookCategories> func_216769_b(RecipeBookContainer<?> p_216769_0_) {
      if (!(p_216769_0_ instanceof WorkbenchContainer) && !(p_216769_0_ instanceof PlayerContainer)) {
         if (p_216769_0_ instanceof FurnaceContainer) {
            return Lists.newArrayList(RecipeBookCategories.FURNACE_SEARCH, RecipeBookCategories.FURNACE_FOOD, RecipeBookCategories.FURNACE_BLOCKS, RecipeBookCategories.FURNACE_MISC);
         } else if (p_216769_0_ instanceof BlastFurnaceContainer) {
            return Lists.newArrayList(RecipeBookCategories.BLAST_FURNACE_SEARCH, RecipeBookCategories.BLAST_FURNACE_BLOCKS, RecipeBookCategories.BLAST_FURNACE_MISC);
         } else {
            return p_216769_0_ instanceof SmokerContainer ? Lists.newArrayList(RecipeBookCategories.SMOKER_SEARCH, RecipeBookCategories.SMOKER_FOOD) : Lists.newArrayList();
         }
      } else {
         return Lists.newArrayList(RecipeBookCategories.SEARCH, RecipeBookCategories.EQUIPMENT, RecipeBookCategories.BUILDING_BLOCKS, RecipeBookCategories.MISC, RecipeBookCategories.REDSTONE);
      }
   }
```
You see that first tab is always SEARCH
I dont see problem with forcing this way, if there is problem with it I remove it

Reply:
@yanny7 I understand that this is the issue here, but my question to clarify:
Does it break if the order is changed: AKA Can I put EQUIPMENT first and SEARCH second?
Obviously this would switch around the tabs, but does it break something else?

Reply:
@OrionDevelopment it breaks anything, but we should stick with some conventions. People would expect that first tab is without filtering, cause all vanilla recipe books works same

Reply:
So pure from a technical standpoint the sorting is not required. It is just a "convention" okey, thanks for clearing that up.